src/index.js

```diff
@@ -196,65 +201,57 @@ export default class Swup {
+		if (this.triggerWillOpenNewWindow(linkEl)) {
+			return;
+		}
+
+		// Exit early if the link should be ignored
+		if (this.options.ignoreLink(linkEl)) {
+			return;
+		}
+
+		// Exit early if control key pressed
+		if (event.metaKey || event.ctrlKey || event.shiftKey || event.altKey) {
+			this.triggerEvent('openPageInNewTab', event);
 			return;
 		}
-		if (this.options.ignoreLink(event.delegateTarget)) {
+
+		// Exit early if other than left mouse button
+		if (event.button !== 0) {
 			return;
 		}
 
-		// no control key pressed
-		if (!event.metaKey && !event.ctrlKey && !event.shiftKey && !event.altKey) {
-			// index of pressed button needs to be checked because Firefox triggers click on all mouse buttons
-			if (event.button === 0) {
-				this.triggerEvent('clickLink', event);
-				event.preventDefault();
-				const link = new Link(event.delegateTarget);
-				if (link.getAddress() == getCurrentUrl() || link.getAddress() == '') {
-					// link to the same URL
-					if (link.getHash() != '') {
-						// link to the same URL with hash
-						this.triggerEvent('samePageWithHash', event);
-						const element = getAnchorElement(link.getHash());
-						if (element != null) {
-							history.replaceState(
-								{
-									url: link.getAddress() + link.getHash(),
-									random: Math.random(),
-									source: 'swup'
-								},
-								document.title,
-								link.getAddress() + link.getHash()
-							);
-						} else {
-							// referenced element not found
-							console.warn(`Element for offset not found (${link.getHash()})`);
-						}
-					} else {
-						// link to the same URL without hash
-						this.triggerEvent('samePage', event);
-					}
-				} else if(!this.isSameResolvedPath(link.getAddress(), getCurrentUrl())) {
-					// link to different url
-					if (link.getHash() != '') {
-						this.scrollToElement = link.getHash();
-					}
-
-					// get custom transition from data
-					let customTransition = event.delegateTarget.getAttribute(
-						'data-swup-transition'
-					);
-
-					// load page
-					this.loadPage(
-						{ url: link.getAddress(), customTransition: customTransition },
-						false
-					);
+		this.triggerEvent('clickLink', event);
+		event.preventDefault();
+
+		const link = new Link(linkEl);
+		const url = link.getAddress();
+		const hash = link.getHash();
+
+		if (url == getCurrentUrl() || url == '') {
+			if (!hash) {
+				// link to the same URL without hash
+				this.triggerEvent('samePage', event);
+			} else {
+				// link to the same URL with hash
+				this.triggerEvent('samePageWithHash', event);
+				const element = getAnchorElement(hash);
+				if (element) {
+					updateHistoryRecord(url + hash);
+				} else {
+					console.warn(`Element for offset not found (#${hash})`);
 				}
 			}
-		} else {
-			// open in new tab (do nothing)
-			this.triggerEvent('openPageInNewTab', event);
+		} else if (!this.isSameResolvedPath(url, getCurrentUrl())) {
```

Comment:
Here would be another chance to do an early returns. If you agree, I could push the change directly to this branch.

Reply:
This one I'm unsure about. Do we exit early here before the first `if` check or after? Maybe you tackle that one.

Reply:
Ok, I'll take care of this.

Reply:
Did some more refactoring. Looking forward to hear what you think! Tests are passing. I also ran `lint` on the code, which turned the option `ignoreLink` to this:

```js
let defaults = {
	//...
	ignoreLink: (el) => 
		el.origin !== window.location.origin || el.closest('[data-no-swup]'),
	//...
};
```
It becomes slightly more readable by an explicit return IMHO:
```js
let defaults = {
	// ...
	ignoreLink: (el) => {
		return el.origin !== window.location.origin || el.closest('[data-no-swup]');
	},
	// ...
};
```

Reply:
The reason I left it unlinted before is for readability. As it is, the one-liner works fine, but once you add more checks, the previous one-check-per-line-format is a lot more readable in my opinion, especially when explaining it in the docs. How do you feel about reverting it and adding a `prettier-ignore` above? Too fussy?

Reply:
> The reason I left it unlinted before is for readability. As it is, the one-liner works fine, but once you add more checks, the previous one-check-per-line-format is a lot more readable in my opinion, especially when explaining it in the docs. How do you feel about reverting it and adding a `prettier-ignore` above? Too fussy?

Hmm, not a fan of adding prettier-ignore there. I would then maybe prefer walking each condition with guard clauses. This will make it VERY clear what's happening. We could even add inline documentation:
```js
let defaults = {
	// ...
	ignoreLink: (el) => {
		// ignore external links:
		if (el.origin !== window.location.origin) return true;
		// ignore links that are inside a container with the attribute [data-no-swup]:
		if (el.closest('[data-no-swup]')) return true;
		// handle all other links
		return false;
	},
	// ...
};
```

Reply:
Yes we shall! :)

Reply:
...ready to merge then!

Reply:
🚀